src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+    while (s <= z[k])
+    {
+      --k;
+      s  = ( (f[q] + q * q) - (f[v[k]] + v[k] * v[k]) ) / (2 * q - 2 * v[k]);
+    }
+
+    k++;
+    v[k] = (double)q;
+    z[k] = s;
+    z[k+1] = +INF;
+  }
+
+  k = 0;
+  for (int q = 0; q <= n-1; q++)
+  {
+    while (z[k+1] < q)
+      k++;
+    d[q] = (q - v[k]) * (q - v[k]) + f[v[k]];
+  }
+  return d;
+}
+
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+dt_2d(MatType& im)
+{
+  arma::vec f(std::max(im.n_rows, im.n_cols));
+  // transform along columns
+  for (size_t x = 0; x < im.n_cols; ++x)
+  {
+    f.subvec(0, im.n_rows - 1) = im.col(x);
+    arma::vec d = this->dt_1d(f, im.n_rows);
+    im.col(x) = d;
+  }
+
+  // transform along rows
+  for (int y = 0; y < im.n_rows; y++)
+  {
+    f.subvec(0, im.n_cols - 1) = im.row(y).t();
+    arma::vec d = this->dt_1d(f, im.n_cols);
+    im.row(y) = d.t();
+  }
+}
+
+/* euclidean distance transform of binary image using squared distance */
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+dt_image(MatType& im, double on)
+{
+  MatType out = MatType(im.n_rows, im.n_cols);
+  out.fill(0.0);
+  out.elem( find(im != on) ).fill(INF);
+  this->dt_2d(out);
+  return out;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+CopyMakeBorder(CubeType& InImage, int top, 
+               int left, int bottom, int right)
+{
+  CubeType OutImage(InImage.n_rows + top + bottom, InImage.n_cols + left + right, InImage.n_slices);
+
+  for(size_t i = 0; i < InImage.n_slices; ++i)
+  {
+    OutImage.slice(i).submat(top, left, InImage.n_rows + top - 1, InImage.n_cols + left - 1)
+     = InImage.slice(i);
+    
+    for(size_t j = 0; j < right; ++j)
+    {
+      OutImage.slice(i).col(InImage.n_cols + left + j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(InImage.n_cols - j - 1);  
+    }
+
+    for(int j = 0; j < left; ++j)
+    {
+      OutImage.slice(i).col(j).subvec(top, InImage.n_rows + top - 1)
+      = InImage.slice(i).col(left - 1 - j);
+    }
+
+    for(int j = 0; j < top; j++)
+    {
+
+      OutImage.slice(i).row(j)
+      = OutImage.slice(i).row(2 * top - 1 - j);
+    }
+     
+    for(int j = 0; j < bottom; j++)
+    {
+      OutImage.slice(i).row(InImage.n_rows + top + j)
+      = OutImage.slice(i).row(InImage.n_rows + top - j - 1);
+    }
+
+  }
+  return OutImage;  
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+RGB2LUV(CubeType& InImage)
+{
+  //assert type is double or float.
+  double a, y0, maxi;
+  a = pow(29.0, 3) / 27.0;
+  y0 = 8.0 / a;
+  maxi = 1.0 / 270.0;
+
+  arma::vec table(1025);
+  for (size_t i = 0; i < 1025; ++i)
+  {
+    table(i) = i / 1024.0;
+    
+    if (table(i) > y0)
+      table(i) = 116 * pow(table(i), 1.0/3.0) - 16.0;
+    else
+      table(i) = table(i) * a;
+
+    table(i) = table(i) * maxi;
+  }
+
```

Comment:
What is the purpose of table?It do not looks like you use it after initialize.Besides, I think we should initialize the table one time only rather than initialize it every time we call this function.